nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp

```diff
@@ -952,7 +952,27 @@ VBufStorage_fieldNode_t* GeckoVBufBackend_t::fillVBuf(
 			// The object has no text, but we do want to render its children.
 			auto [varChildren, accChildRes] = getAccessibleChildren(pacc, 0, childCount);
 			if (S_OK != accChildRes || varChildren.size() == 0) {
-				LOG_ERROR(L"AccessibleChildren failed (count: " << childCount << L"), res: " << accChildRes);
+				std::wstringstream msg;
+				msg << L"AccessibleChildren failed (count: " << childCount << L"), res: " << accChildRes;
+				switch (accChildRes) {
+				case E_NOINTERFACE:
+					msg << L" (E_NOINTERFACE, No such interface supported)";
+					// This is certainly an error, and indicates a bug in the IA2 provider.
+					LOG_ERROR(msg.str());
+					break;
+				case CO_E_OBJNOTCONNECTED:
+					msg << L" (CO_E_OBJNOTCONNECTED, Object is not connected to server)";
+					// CO_E_OBJNOTCONNECTED indicates that the parent died since the query
+					// to accChildCount, and returning no children indicates that all children

```

Comment:
Returning no children isn't covered by this case. In that case, varChildren.size() == 0 will be true and the result will be S_FALSE. You'll need to cover that separately.

Reply:
Oh I see, it's the other kind of "and", I misread that the first time, good catch.

Reply:
This case `(S_FALSE, no children)` sounds like it could be specific to Firefox. I wouldn't expect any children returned in any other case with `S_FALSE`, how do we differentiate this situation from some "generic failure"?

Reply:
I don't quite follow the reasoning here. From the [AccessibleChildren documentation](https://docs.microsoft.com/en-us/windows/win32/api/oleacc/nf-oleacc-accessiblechildren):

> S_FALSE
> The function succeeded, but there are fewer elements in the rgvarChildren array than there are children requested in cChildren.

So, if you request 1 child and you get 0 back (in this case, because one of them died), you'll get S_FALSE.

A generic failure should return an E_XXX error, not S_FALSE. S_FALSE is [explicitly not a failure](https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com):

> The success code S_FALSE deserves mention. Some methods use S_FALSE to mean, roughly, a negative condition that is not a failure. It can also indicate a "no-op"—the method succeeded, but had no effect.

Reply:
Ah looks like I conflated `S_FALSE` and `E_FAIL`, that and most code paths only check for `S_OK` treating anything else as a failure.

I'll take another pass at this.

Reply:
For reference, Windows has the SUCCEEDED and FAILED macros, the former checking for S_OK and S_FALSE and the latter checking for E_XXX.

Most of the time in this code, it's not so much that we treat S_FALSE as failure, but more that we want no data/no-op and failure to be handled the same way; i.e. we just treat failure the same as no data/no-op. The intention was to be as fault tolerant as possible.